corgi/web/templates/provided_relationships_manifest.json

```diff
@@ -0,0 +1,11 @@
+{% for node in obj.get_provides_nodes %}
+    {
+      "relatedSpdxElement": "SPDXRef-{{obj.uuid}}",{# subcomponent is built from, or contained in, component #}
+      "relationshipType": {% if node.type == "PROVIDES_DEV" %}"DEV_DEPENDENCY_OF"{% else %}"CONTAINED_BY"{% endif %},
+      "spdxElementId": "SPDXRef-{{node.obj.uuid}}"
```

Comment:
The specification and documentation is not very clear here, but I think that relatedSpdxElement is on the right side of the relationship when it's all in one line. I think that because following [RDF example](https://spdx.github.io/spdx-spec/v2.2.2/relationships-between-SPDX-elements/#111-relationship-field) which embeds a relationship inside a file element:

```
<File rdf:about="#SPDXRef-45">
  <relationship>
    <Relationship>
      <relatedSpdxElement>
        <File rdf:about="http://spdx.org/spdxdocs/spdx-tools-v1.2-3F2504E0-4F89-41D3-9A0C-0305E82..."/>
      </relatedSpdxElement>
      <relationshipType>http://spdx.org/rdf/terms#relationshipType_contains
      </relationshipType>
    </Relationship>
  </relationship>
```

I read this as `SPDXRef-45 CONTAINS spdx-tools-v1.2`. Notice the `spdx-tools-v1.2` has the 'relatedSpdxElement' label here, the same as the JSON element above.

Given that, I think I have the relationship around the right way. The [spec ](https://spdx.github.io/spdx-spec/v2.2.2/relationships-between-SPDX-elements/#111-relationship-field) expands on the 'DEV_DEPENCY_OF relationship by saying "Is to be used when SPDXRef-A is a development dependency of SPDXRef-B." 

Here's an example of an actualy PROVIDES_DEV relationship from one of our containers [acm-governance-policy-addon-controller-container](https://brewweb.engineering.redhat.com/brew/buildinfo?buildID=2157159) produced with the existing code.

```
    {
      relatedSpdxElement: SPDXRef-38049df5-0dd3-4859-a450-3ab63e2a4b65,
      relationshipType: DEV_DEPENDENCY_OF,
      spdxElementId: SPDXRef-d291152d-7a4e-49d0-9249-a49a7f686961
    },
```

Expanding the SpdxIds:
```
$ cat manifest.1 | jq '.packages[] | select(.SPDXID=="SPDXRef-38049df5-0dd3-4859-a450-3ab63e2a4b65") | .externalRefs'
[
  {
    "referenceCategory": "PACKAGE_MANAGER",
    "referenceLocator": "pkg:oci/search-api-container@sha256:cee5135056464f3ea5d08cd12a8eb49b48de29b9f82ca62a2a4aced9e9823bba?tag=v2.6.1-1",
    "referenceType": "purl"
  }
]
cat manifest.1 | jq '.packages[] | select(.SPDXID=="SPDXRef-d291152d-7a4e-49d0-9249-a49a7f686961") | .externalRefs'
[
  {
    "referenceCategory": "PACKAGE_MANAGER",
    "referenceLocator": "pkg:npm/whatwg-encoding@1.0.5",
    "referenceType": "purl"
  }
]
```

It reads: `npm/whatwg-encoding DEV_DEPENDENCY_OF search-api-container` which is correct.
```
    {
      relatedSpdxElement: search-api-container,
      relationshipType: DEV_DEPENDENCY_OF,
      spdxElementId: npm/whatwg-encoding
    },
```

Just another [example](https://github.com/spdx/tools-java/blob/92e1c5d29eacb0081139b0c05e5e6270b231788c/testResources/SPDXJSONExample-v2.2.spdx.json#L252-L256) to show the relatedSpdxElement should be on the right hand side:
```
{
    "spdxElementId" : "SPDXRef-DOCUMENT",
    "relatedSpdxElement" : "SPDXRef-Package",
    "relationshipType" : "DESCRIBES"
  }
```



You can read more about the PROVIDES_DEV relationship we get from [OSBS here](https://github.com/containerbuildsystem/cachito/#feature-definitions).

Reply:
tl;dr - you're right, I am very confused. There may be a bug in our docs / relationship types, or maybe I'm still confused. Maybe we should use BUILD_DEPENDENCY_OF, otherwise code you have here LGTM.

> The specification and documentation is not very clear here, but I think that relatedSpdxElement is on the right side of the relationship when it's all in one line. 
> I read this as SPDXRef-45 CONTAINS spdx-tools-v1.2. Notice the spdx-tools-v1.2 has the 'relatedSpdxElement' label here, the same as the JSON element above.

Yep, I'm reading this the same way. I think my confusion was the difference between PROVIDES (contains / contained_by) and PROVIDES_DEV. For the first, I originally said ```container CONTAINS component```, so the provided object is "below" ithe container in the hierarchy. From the container, you get the component.

But for PROVIDES_DEV, we'd say ```component is a DEPENDENCY_OF container``` that's needed to build the container, so the provided object is "above" the container in the hierarchy. From the component, you get the container.

After thinking through this a bit too long, what you have (using CONTAINED_BY instead of CONTAINS and flipping the ordering from my original version) seems right. The provided component is always on the left side, and RELATES_TO the component being manifested on the right side, either as a DEPENDENCY_OF or something CONTAINED_BY something else. Thanks for spending the time to generate those examples, and for catching my mistake :)

The info in docs/user_guide.md may also need to be updated:
```
`provides`: ....a list of components that this component provides,
   ....as components built from this component
```
(Below is considering only PROVIDES_DEV)
For type=rpm, this relationship makes sense since the SRPM "provides" the RPMs. AKA the RPMs are built from the SRPM, and the SRPM is a BUILD_DEPENDENCY_OF the RPMs.

But for type=container_image, isn't it the other way around? The container "provides" the RPMs, but the container is built from the RPMs, so each RPM should be a BUILD_DEPENDENCY_OF the container. The RPMs aren't built from the container, as the docs imply.

Do we need another relationship type to distinguish this? Otherwise we'll have one relationship type that can be interpreted two different ways, depending on if you're talking about RPMs or containers.

Reply:
I think the user_guide for "provides" is OK. The relationship type itself could be split out into more types, but it seems like overkill for the complexity it will create.

OSBS doesn't distinguish between build and dev dependencies the way SPDX does.